# Pandas Getting Started

-> Pandas is used to analyze, clean, exploring and manipulating data .

## Pandas Installation

In [1]:
pip install pandas

Note: you may need to restart the kernel to use updated packages.


## Importing Pandas

In [2]:
import pandas as pd

###  Checking the pandas version

In [3]:
print(pd.__version__)

1.4.4


## Pandas Series 

In [4]:
num = [2, 3, 5, 7] # list of numbers

series = pd.Series(num) # pandas series

print(series)

0    2
1    3
2    5
3    7
dtype: int64


###  Labelling

-> By default pandas series are indexed

In [5]:
series_1 = pd.Series(num, index = ['A', 'B', 'C', 'D'])

print(series_1)

A    2
B    3
C    5
D    7
dtype: int64


### Accessing Elements Using Label

In [6]:
print(series_1['B']) #prints the second element in series_1 variable.

3


###  Key/values object as Series

In [7]:
days = {'sunday': 1, 'Monday': 2, 'tuesday': 3, 'wednesday': 4, 'thursday': 5, 'friday': 6, 'saturday': 7}
day_series = pd.Series(days)

print(days)

{'sunday': 1, 'Monday': 2, 'tuesday': 3, 'wednesday': 4, 'thursday': 5, 'friday': 6, 'saturday': 7}


##  Dataframes

-> Data sets in pandas are usually multi-dimensional table

In [8]:
data = {'Names': ['Kevin', 'Brian', 'Winney', 'Violet'],
       'Ages': [32, 21, 18, 20]}

data_set = pd.DataFrame(data, index = [1, 2, 3, 4]) # loading data into a DataFrame object.

print(data_set)

    Names  Ages
1   Kevin    32
2   Brian    21
3  Winney    18
4  Violet    20


### Locating Row

-> A DataFrame is like a table with both columns and rows

#### loc

-> is pandas function that return one or more specified row(s)

In [9]:
print(data_set.loc[1])

Names    Kevin
Ages        32
Name: 1, dtype: object


In [10]:
print(data_set.loc[[1, 2]])

   Names  Ages
1  Kevin    32
2  Brian    21


## Loading Files into a DataFrame

In [11]:
df = pd.read_csv('Iris.csv')

print(df)

        CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD  TAX  \
0    0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900    1  296   
1    0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671    2  242   
2    0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671    2  242   
3    0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622    3  222   
4    0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622    3  222   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...  ...   
501  0.06263   0.0  11.93   0.0  0.573  6.593  69.1  2.4786    1  273   
502  0.04527   0.0  11.93   0.0  0.573  6.120  76.7  2.2875    1  273   
503  0.06076   0.0  11.93   0.0  0.573  6.976  91.0  2.1675    1  273   
504  0.10959   0.0  11.93   0.0  0.573  6.794  89.3  2.3889    1  273   
505  0.04741   0.0  11.93   0.0  0.573  6.030   NaN  2.5050    1  273   

     PTRATIO       B  LSTAT  MEDV  
0       15.3  396.90   4.98  24.0  
1       17.8  396.90   9.14  21.6  
2       17.8  3

## Read CSV Files

-> A simple way to store big data set is to use comma separated values files(CSV)


In [12]:
print(df.to_string()) # to_string() is used to print entire DataFrame.

         CRIM     ZN  INDUS  CHAS     NOX     RM    AGE      DIS  RAD  TAX  PTRATIO       B  LSTAT  MEDV
0     0.00632   18.0   2.31   0.0  0.5380  6.575   65.2   4.0900    1  296     15.3  396.90   4.98  24.0
1     0.02731    0.0   7.07   0.0  0.4690  6.421   78.9   4.9671    2  242     17.8  396.90   9.14  21.6
2     0.02729    0.0   7.07   0.0  0.4690  7.185   61.1   4.9671    2  242     17.8  392.83   4.03  34.7
3     0.03237    0.0   2.18   0.0  0.4580  6.998   45.8   6.0622    3  222     18.7  394.63   2.94  33.4
4     0.06905    0.0   2.18   0.0  0.4580  7.147   54.2   6.0622    3  222     18.7  396.90    NaN  36.2
5     0.02985    0.0   2.18   0.0  0.4580  6.430   58.7   6.0622    3  222     18.7  394.12   5.21  28.7
6     0.08829   12.5   7.87   NaN  0.5240  6.012   66.6   5.5605    5  311     15.2  395.60  12.43  22.9
7     0.14455   12.5   7.87   0.0  0.5240  6.172   96.1   5.9505    5  311     15.2  396.90  19.15  27.1
8     0.21124   12.5   7.87   0.0  0.5240  5.631  100.0

###  Max_rows

-> The number of rows returned is defined in optional settings

-> You can check your system's maximum_rows with the:
  pd.options.display.max.rows statement

In [13]:
print(pd.options.display.max_rows)

60


###  Increasing the number of rows to display the entire DataFrame

In [14]:
pd.options.display.max_rows = 9999

df = pd.read_csv('Iris.csv')
df

CRIM     ZN  INDUS  CHAS     NOX     RM    AGE      DIS  RAD  TAX  \
0     0.00632   18.0   2.31   0.0  0.5380  6.575   65.2   4.0900    1  296   
1     0.02731    0.0   7.07   0.0  0.4690  6.421   78.9   4.9671    2  242   
2     0.02729    0.0   7.07   0.0  0.4690  7.185   61.1   4.9671    2  242   
3     0.03237    0.0   2.18   0.0  0.4580  6.998   45.8   6.0622    3  222   
4     0.06905    0.0   2.18   0.0  0.4580  7.147   54.2   6.0622    3  222   
5     0.02985    0.0   2.18   0.0  0.4580  6.430   58.7   6.0622    3  222   
6     0.08829   12.5   7.87   NaN  0.5240  6.012   66.6   5.5605    5  311   
7     0.14455   12.5   7.87   0.0  0.5240  6.172   96.1   5.9505    5  311   
8     0.21124   12.5   7.87   0.0  0.5240  5.631  100.0   6.0821    5  311   
9     0.17004   12.5   7.87   NaN  0.5240  6.004   85.9   6.5921    5  311   
10    0.22489   12.5   7.87   0.0  0.5240  6.377   94.3   6.3467    5  311   
11    0.11747   12.5   7.87   0.0  0.5240  6.009   82.9   6.2267    5  311   
12    0.09378   12.5   7.87   0.0  0.5240  5.889   39.0   5.4509    5  311   
13    0.62976    0.0   8.14   0.0  0.5380  5.949   61.8   4.7075    4  307   
14    0.63796    0.0   8.14   NaN  0.5380  6.096   84.5   4.4619    4  307   
15    0.62739    0.0   8.14   0.0  0.5380  5.834   56.5   4.4986    4  307   
16    1.05393    0.0   8.14   0.0  0.5380  5.935   29.3   4.4986    4  307   
17    0.78420    0.0   8.14   0.0  0.5380  5.990   81.7   4.2579    4  307   
18    0.80271    0.0   8.14   0.0  0.5380  5.456   36.6   3.7965    4  307   
19    0.72580    0.0   8.14   0.0  0.5380  5.727   69.5   3.7965    4  307   
20    1.25179    0.0   8.14   0.0  0.5380  5.570   98.1   3.7979    4  307   
21    0.85204    0.0   8.14   0.0  0.5380  5.965   89.2   4.0123    4  307   
22    1.23247    0.0   8.14   0.0  0.5380  6.142   91.7   3.9769    4  307   
23    0.98843    0.0   8.14   0.0  0.5380  5.813  100.0   4.0952    4  307   
24    0.75026    0.0   8.14   0.0  0.5380  5.924   94.1   4.3996    4  307   
25    0.84054    0.0   8.14   0.0  0.5380  5.599   85.7   4.4546    4  307   
26    0.67191    0.0   8.14   0.0  0.5380  5.813   90.3   4.6820    4  307   
27    0.95577    0.0   8.14   0.0  0.5380  6.047   88.8   4.4534    4  307   
28    0.77299    0.0   8.14   0.0  0.5380  6.495   94.4   4.4547    4  307   
29    1.00245    0.0   8.14   0.0  0.5380  6.674   87.3   4.2390    4  307   
30    1.13081    0.0   8.14   0.0  0.5380  5.713   94.1   4.2330    4  307   
31    1.35472    0.0   8.14   0.0  0.5380  6.072  100.0   4.1750    4  307   
32    1.38799    0.0   8.14   0.0  0.5380  5.950   82.0   3.9900    4  307   
33    1.15172    0.0   8.14   0.0  0.5380  5.701   95.0   3.7872    4  307   
34    1.61282    0.0   8.14   0.0  0.5380  6.096   96.9   3.7598    4  307   
35    0.06417    0.0   5.96   0.0  0.4990  5.933   68.2   3.3603    5  279   
36    0.09744    0.0    NaN   0.0  0.4990  5.841   61.4   3.3779    5  279   
37    0.08014    0.0   5.96   0.0  0.4990  5.850   41.5   3.9342    5  279   
38    0.17505    0.0   5.96   0.0  0.4990  5.966   30.2   3.8473    5  279   
39    0.02763   75.0   2.95   0.0  0.4280  6.595   21.8   5.4011    3  252   
40    0.03359   75.0   2.95   0.0  0.4280  7.024   15.8   5.4011    3  252   
41    0.12744    0.0   6.91   0.0  0.4480  6.770    2.9   5.7209    3  233   
42    0.14150    0.0   6.91   0.0  0.4480  6.169    6.6   5.7209    3  233   
43    0.15936    0.0   6.91   NaN  0.4480  6.211    6.5   5.7209    3  233   
44    0.12269    0.0   6.91   0.0  0.4480  6.069   40.0   5.7209    3  233   
45    0.17142    0.0   6.91   0.0  0.4480  5.682   33.8   5.1004    3  233   
46    0.18836    0.0   6.91   0.0  0.4480  5.786   33.3   5.1004    3  233   
47    0.22927    0.0    NaN   0.0  0.4480  6.030   85.5   5.6894    3  233   
48    0.25387    0.0   6.91   0.0  0.4480  5.399   95.3   5.8700    3  233   
49    0.21977    0.0   6.91   0.0  0.4480  5.602   62.0   6.0877    3  233   
50    0.08873   21.0   5.64   0

#  Analyzing DataFrame

##  Viewing Data

#### head()

-> head() function returns the first five rows of the Dataframe by default unless specified

In [15]:
print(df.head())

      CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD  TAX  PTRATIO  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900    1  296     15.3   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671    2  242     17.8   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671    2  242     17.8   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622    3  222     18.7   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622    3  222     18.7   

        B  LSTAT  MEDV  
0  396.90   4.98  24.0  
1  396.90   9.14  21.6  
2  392.83   4.03  34.7  
3  394.63   2.94  33.4  
4  396.90    NaN  36.2  


In [16]:
print(df.head().to_string()) # print the entire dataframe in a row

      CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD  TAX  PTRATIO       B  LSTAT  MEDV
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900    1  296     15.3  396.90   4.98  24.0
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671    2  242     17.8  396.90   9.14  21.6
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671    2  242     17.8  392.83   4.03  34.7
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622    3  222     18.7  394.63   2.94  33.4
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622    3  222     18.7  396.90    NaN  36.2


####  tail()

-> prints the last five rows of the DataFrame by default unless specified

In [17]:
print(df.tail(10))

        CRIM   ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD  TAX  PTRATIO  \
496  0.28960  0.0   9.69   0.0  0.585  5.390  72.9  2.7986    6  391     19.2   
497  0.26838  0.0   9.69   0.0  0.585  5.794  70.6  2.8927    6  391     19.2   
498  0.23912  0.0   9.69   0.0  0.585  6.019  65.3  2.4091    6  391     19.2   
499  0.17783  0.0   9.69   0.0  0.585  5.569  73.5  2.3999    6  391     19.2   
500  0.22438  0.0   9.69   0.0  0.585  6.027  79.7  2.4982    6  391     19.2   
501  0.06263  0.0  11.93   0.0  0.573  6.593  69.1  2.4786    1  273     21.0   
502  0.04527  0.0  11.93   0.0  0.573  6.120  76.7  2.2875    1  273     21.0   
503  0.06076  0.0  11.93   0.0  0.573  6.976  91.0  2.1675    1  273     21.0   
504  0.10959  0.0  11.93   0.0  0.573  6.794  89.3  2.3889    1  273     21.0   
505  0.04741  0.0  11.93   0.0  0.573  6.030   NaN  2.5050    1  273     21.0   

          B  LSTAT  MEDV  
496  396.90  21.14  19.7  
497  396.90  14.10  18.3  
498  396.90  12.92  21.2  


In [18]:
print(df.tail().to_string())

        CRIM   ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD  TAX  PTRATIO       B  LSTAT  MEDV
501  0.06263  0.0  11.93   0.0  0.573  6.593  69.1  2.4786    1  273     21.0  391.99    NaN  22.4
502  0.04527  0.0  11.93   0.0  0.573  6.120  76.7  2.2875    1  273     21.0  396.90   9.08  20.6
503  0.06076  0.0  11.93   0.0  0.573  6.976  91.0  2.1675    1  273     21.0  396.90   5.64  23.9
504  0.10959  0.0  11.93   0.0  0.573  6.794  89.3  2.3889    1  273     21.0  393.45   6.48  22.0
505  0.04741  0.0  11.93   0.0  0.573  6.030   NaN  2.5050    1  273     21.0  396.90   7.88  11.9


#### info()

-> give the information about the data

In [19]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     486 non-null    float64
 1   ZN       486 non-null    float64
 2   INDUS    486 non-null    float64
 3   CHAS     486 non-null    float64
 4   NOX      506 non-null    float64
 5   RM       506 non-null    float64
 6   AGE      486 non-null    float64
 7   DIS      506 non-null    float64
 8   RAD      506 non-null    int64  
 9   TAX      506 non-null    int64  
 10  PTRATIO  506 non-null    float64
 11  B        506 non-null    float64
 12  LSTAT    486 non-null    float64
 13  MEDV     506 non-null    float64
dtypes: float64(12), int64(2)
memory usage: 55.5 KB
None


-> in above info() also give the  number of non-null values present in each columns.


-> In our data it seems there are 486 out of 506 in 'CRIM', 'ZN' etc.

-> This means there are 20 rows in those columns with no values

-> Empty values can be bad when analyzing data, and you should consider removing them.


## Data Cleaning in Pandas

-> Data cleaning means fixing bad data in your data set.

-> Bad data could be:
                    
                    *Empty cells
                    
                     *Data in wrong format
                    
                     *Wrong data
                    
                     *Duplicates

###  Empty cells

-> Empty cells can potentially give you a wrong result when you analyze data.

-> One way to deal with the null row is to drop their values

### Checking for null values

#### isna() / isnull()

-> isna() function is used to check for null values in our data

In [20]:
print(df.isna()) # this return boolean true for non-empty row and false values null rows 

      CRIM     ZN  INDUS   CHAS    NOX     RM    AGE    DIS    RAD    TAX  \
0    False  False  False  False  False  False  False  False  False  False   
1    False  False  False  False  False  False  False  False  False  False   
2    False  False  False  False  False  False  False  False  False  False   
3    False  False  False  False  False  False  False  False  False  False   
4    False  False  False  False  False  False  False  False  False  False   
5    False  False  False  False  False  False  False  False  False  False   
6    False  False  False   True  False  False  False  False  False  False   
7    False  False  False  False  False  False  False  False  False  False   
8    False  False  False  False  False  False  False  False  False  False   
9    False  False  False   True  False  False  False  False  False  False   
10   False  False  False  False  False  False  False  False  False  False   
11   False  False  False  False  False  False  False  False  False  False   

In [21]:
print(df.isna().sum() # returns the total number of empty rows in each column

SyntaxError: unexpected EOF while parsing (224084434.py, line 1)

###  Dropping  Null values

#### dropna()

-> dropna() is used to drop null values in a dataset.

-> Howerver, for this case I will not drop the null values but I comment the code.

In [ ]:
# Print(df.dropna(inplace = True)) this will drop the null values


### Replacing the null values

-> This is another way of dealing with a empty values.

-> Is considered the best option rather than dropping the value

#### fillna() 

### Replacing with specified value

-> fillna() function is used to fill/ replaces the null values with other values

In [58]:
df['CRIM'].fillna(0.324011, inplace = True) 


#this will replace empty cells with 0.324011 in 'CRIM' column.

print(df)


          CRIM          ZN  INDUS       CHAS     NOX     RM         AGE  \
0     0.006320   18.000000   2.31   0.000000  0.5380  6.575   65.200000   
1     0.027310    0.000000   7.07   0.000000  0.4690  6.421   78.900000   
2     0.027290    0.000000   7.07   0.000000  0.4690  7.185   61.100000   
3     0.032370    0.000000   2.18   0.000000  0.4580  6.998   45.800000   
4     0.069050    0.000000   2.18   0.000000  0.4580  7.147   54.200000   
5     0.029850    0.000000   2.18   0.000000  0.4580  6.430   58.700000   
6     0.088290   12.500000   7.87  11.211934  0.5240  6.012   66.600000   
7     0.144550   12.500000   7.87   0.000000  0.5240  6.172   96.100000   
8     0.211240   12.500000   7.87   0.000000  0.5240  5.631  100.000000   
9     0.170040   12.500000   7.87  11.211934  0.5240  6.004   85.900000   
10    0.224890   12.500000   7.87   0.000000  0.5240  6.377   94.300000   
11    0.117470   12.500000   7.87   0.000000  0.5240  6.009   82.900000   
12    0.093780   12.50000

In [33]:
df['CRIM'].isna().sum() 

# this confirms that all empty cell are replaced in 'CRIM' column

0

### Replacing using mean(), median() and mode()

-> This is the common way used by most analysts

###  Replacing with mean

In [52]:
mean = df['ZN'].mean() #mean of ZN column

print(mean)

11.211934156378598


### I'll replace null values with mean of  'ZN' column

In [55]:
df.fillna(mean, inplace = True)

print(df)



          CRIM          ZN  INDUS       CHAS     NOX     RM         AGE  \
0     0.006320   18.000000   2.31   0.000000  0.5380  6.575   65.200000   
1     0.027310    0.000000   7.07   0.000000  0.4690  6.421   78.900000   
2     0.027290    0.000000   7.07   0.000000  0.4690  7.185   61.100000   
3     0.032370    0.000000   2.18   0.000000  0.4580  6.998   45.800000   
4     0.069050    0.000000   2.18   0.000000  0.4580  7.147   54.200000   
5     0.029850    0.000000   2.18   0.000000  0.4580  6.430   58.700000   
6     0.088290   12.500000   7.87  11.211934  0.5240  6.012   66.600000   
7     0.144550   12.500000   7.87   0.000000  0.5240  6.172   96.100000   
8     0.211240   12.500000   7.87   0.000000  0.5240  5.631  100.000000   
9     0.170040   12.500000   7.87  11.211934  0.5240  6.004   85.900000   
10    0.224890   12.500000   7.87   0.000000  0.5240  6.377   94.300000   
11    0.117470   12.500000   7.87   0.000000  0.5240  6.009   82.900000   
12    0.093780   12.50000

In [59]:


df.isna().sum() 

#this confirms that the empty values are all filled.

CRIM       0
ZN         0
INDUS      0
CHAS       0
NOX        0
RM         0
AGE        0
DIS        0
RAD        0
TAX        0
PTRATIO    0
B          0
LSTAT      0
MEDV       0
dtype: int64

### Handling Wrong Data Format 

-> Cells with data of wrong format can make it difficult, or even impossible, to analyze data.


->To fix it, you have two options:

                                1.remove the rows
                                
                                2.convert all cells in the columns into the same format.

### Handling Duplicates

-> Duplicates are those rows/ values registered more than one time.

####  Checking for Duplicates

In [63]:
df.duplicated().sum()

# duplicate() function helps to check for duplicate values

0

###  Dropping Duplicates

-> drop_duplicates() method is used to remove the duplicates in the data

In [64]:
#there are no duplicates so there is no need to remove any

### Data Correlations

-> A great aspect of the Pandas module is the corr() method.


->The corr() method calculates the relationship between each column in your data set.

In [67]:
df.corr() #this shows the relationship between the columns

CRIM        ZN     INDUS      CHAS       NOX        RM       AGE  \
CRIM     1.000000 -0.185840  0.392640  0.021476  0.410983 -0.219962  0.336034   
ZN      -0.185840  1.000000 -0.513079 -0.033723 -0.502287  0.316550 -0.492424   
INDUS    0.392640 -0.513079  1.000000  0.075103  0.738387 -0.377978  0.545182   
CHAS     0.021476 -0.033723  0.075103  1.000000  0.050521 -0.073423  0.048459   
NOX      0.410983 -0.502287  0.738387  0.050521  1.000000 -0.302188  0.648420   
RM      -0.219962  0.316550 -0.377978 -0.073423 -0.302188  1.000000 -0.232608   
AGE      0.336034 -0.492424  0.545182  0.048459  0.648420 -0.232608  1.000000   
DIS     -0.366058  0.638388 -0.698621  0.003636 -0.769230  0.205246 -0.662799   
RAD      0.601420 -0.306316  0.592735  0.087056  0.611441 -0.209847  0.430374   
TAX      0.560629 -0.308334  0.716267  0.076429  0.668023 -0.292048  0.472053   
PTRATIO  0.277885 -0.403085  0.385366 -0.006964  0.188933 -0.355501  0.246163   
B       -0.365459  0.167431 -0.354840 -0.005720 -0.380051  0.128069 -0.225104   
LSTAT    0.437235 -0.405102  0.564513  0.025514  0.573080 -0.604477  0.537878   
MEDV    -0.383830  0.365943 -0.476394 -0.056581 -0.427321  0.695360 -0.364935   

              DIS       RAD       TAX   PTRATIO         B     LSTAT      MEDV  
CRIM    -0.366058  0.601420  0.560629  0.277885 -0.365459  0.437235 -0.383830  
ZN       0.638388 -0.306316 -0.308334 -0.403085  0.167431 -0.405102  0.365943  
INDUS   -0.698621  0.592735  0.716267  0.385366 -0.354840  0.564513 -0.476394  
CHAS     0.003636  0.087056  0.076429 -0.006964 -0.005720  0.025514 -0.056581  
NOX     -0.769230  0.611441  0.668023  0.188933 -0.380051  0.573080 -0.427321  
RM       0.205246 -0.209847 -0.292048 -0.355501  0.128069 -0.604477  0.695360  
AGE     -0.662799  0.430374  0.472053  0.246163 -0.225104  0.537878 -0.364935  
DIS      1.000000 -0.494588 -0.534432 -0.232471  0.291512 -0.483152  0.249929  
RAD     -0.494588  1.000000  0.910228  0.464741 -0.444413  0.467592 -0.381626  
TAX     -0.534432  0.910228  1.000000  0.460853 -0.441808  0.524023 -0.468536  
PTRATIO -0.232471  0.464741  0.460853  1.000000 -0.177383  0.370237 -0.507787  
B        0.291512 -0.444413 -0.441808 -0.177383  1.000000 -0.371303  0.333461  
LSTAT   -0.483152  0.467592  0.524023  0.370237 -0.371303  1.000000 -0.723191  
MEDV     0.249929 -0.381626 -0.468536 -0.507787  0.333461 -0.723191  1.000000